gitops/snapshot_integration_tests_status.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions andF
+limitations under the License.
+*/
+
+package gitops
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/api/v1beta1"
+	"github.com/redhat-appstudio/operator-toolkit/metadata"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const integrationTestStatusesSchema = `{
+	"$schema": "http://json-schema.org/draft/2020-12/schema#",
+	"type":  "array",
+	"items": {
+	  "type": "object",
+      "properties": {
+        "scenario": {
+          "type": "string"
+        },
+        "status": {
+          "type": "string"
+        },
+        "lastUpdateTime": {
+          "type": "string"
+        },
+        "details": {
+          "type": "string"
+        },
+        "startTime": {
+          "type": "string"
+        },
+        "completionTime": {
+          "type": "string"
+        }
+      },
+	  "required": ["scenario", "status", "lastUpdateTime"]
+	}
+  }`
+
+// IntegrationTestStatusDetail contains metadata about the particular scenario testing status
+type IntegrationTestStatusDetail struct {
+	// ScenarioName name
+	ScenarioName string `json:"scenario"`
+	// The status summary for the ITS and Snapshot
+	Status IntegrationTestStatus `json:"status"`
+	// The time of reporting the status
+	LastUpdateTime time.Time `json:"lastUpdateTime"`
+	// The details of reported status
+	Details string `json:"details"`
+	// Startime when we moved to inProgress
+	StartTime *time.Time `json:"startTime,omitempty"` // pointer to make omitempty work
+	// Completion time when test failed or passed
+	CompletionTime *time.Time `json:"completionTime,omitempty"` // pointer to make omitempty work
+}
+
+// SnapshotIntegrationTestStatuses type handles details about snapshot tests
+// Please note that internal representation differs from marshalled representation
+// Data are not written directly into snapshot, they are just cached in this structure
+type SnapshotIntegrationTestStatuses struct {
+	// map scenario name to test details
+	statuses map[string]*IntegrationTestStatusDetail
+	// flag if any updates have been done
+	dirty bool
```

Comment:
I assume this is k8s convention, but why does it matter if the status has changed?

Reply:
if status doesn't change, It won't be trying to update object

Reply:
Got it, thanks for the info